### PR TITLE
fix: java.lang.IndexOutOfBoundsException when trying to "insert @RestClient"

### LINKED ISSUE #764: java.lang.IndexOutOfBoundsException when trying to "insert @RestClient"
**Steps:**
1. EXEC: have git@github.com:adietish/code-with-quarkus.git cloned and imported into IJ
2. EXEC: open `ExtenstionsResource.java` in an editor
3. EXEC: put cursor into `extensionsService` and activate the suggestions (press Alt + Return)
4. ASSERT: "Insert @RestClient" is suggested
image
6. EXEC: select it

**Result:**
"@RestClient" is not inserted, fatal IDE error is displayed with the following stacktrace:
```
java.lang.IndexOutOfBoundsException: Wrong line: 22. Available lines count: 21
	at com.intellij.openapi.editor.impl.LineSet.checkLineIndex(LineSet.java:200)
	at com.intellij.openapi.editor.impl.LineSet.getLineStart(LineSet.java:181)
	at com.intellij.openapi.editor.impl.DocumentImpl.getLineStartOffset(DocumentImpl.java:1027)
	at com.redhat.devtools.intellij.quarkus.lsp4ij.LSPIJUtils.toOffset(LSPIJUtils.java:120)
	at com.redhat.devtools.intellij.quarkus.lsp4ij.LSPIJUtils.applyWorkspaceEdit(LSPIJUtils.java:219)
	at com.redhat.devtools.intellij.quarkus.lsp4ij.LSPIJUtils.applyWorkspaceEdit(LSPIJUtils.java:157)
	at com.redhat.devtools.intellij.quarkus.lsp4ij.operations.codeactions.LSPCodeActionIntentionAction.apply(LSPCodeActionIntentionAction.java:96)
	at com.redhat.devtools.intellij.quarkus.lsp4ij.operations.codeactions.LSPCodeActionIntentionAction.invoke(LSPCodeActionIntentionAction.java:84)
	at com.intellij.codeInsight.intention.impl.IntentionActionWithTextCaching$MyIntentionAction.lambda$invoke$0(IntentionActionWithTextCaching.java:223)
	at com.intellij.util.SlowOperations.allowSlowOperations(SlowOperations.java:194)
	at com.intellij.codeInsight.intention.impl.IntentionActionWithTextCaching$MyIntentionAction.invoke(IntentionActionWithTextCaching.java:223)
	at com.intellij.codeInsight.intention.impl.ShowIntentionActionsHandler.lambda$invokeIntention$8(ShowIntentionActionsHandler.java:273)
	at com.intellij.openapi.application.WriteAction.lambda$run$1(WriteAction.java:87)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteActionWithClass(ApplicationImpl.java:946)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteAction(ApplicationImpl.java:972)
	at com.intellij.openapi.application.WriteAction.run(WriteAction.java:86)
	at com.intellij.codeInsight.intention.impl.ShowIntentionActionsHandler.invokeIntention(ShowIntentionActionsHandler.java:273)
	at com.intellij.codeInsight.intention.impl.ShowIntentionActionsHandler.lambda$chooseActionAndInvoke$7(ShowIntentionActionsHandler.java:250)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:219)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:174)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:164)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:150)
	at com.intellij.codeInsight.intention.impl.ShowIntentionActionsHandler.chooseActionAndInvoke(ShowIntentionActionsHandler.java:249)
	at com.intellij.codeInsight.intention.impl.IntentionListStep.chooseActionAndInvoke(IntentionListStep.java:137)
	at com.intellij.codeInsight.intention.impl.IntentionListStep.lambda$applyAction$0(IntentionListStep.java:129)
	at com.intellij.ui.popup.AbstractPopup.lambda$dispose$17(AbstractPopup.java:1618)
	at com.intellij.openapi.wm.impl.FocusManagerImpl.lambda$doWhenFocusSettlesDown$3(FocusManagerImpl.java:172)
	at com.intellij.util.ui.EdtInvocationManager.invokeLaterIfNeeded(EdtInvocationManager.java:33)
	at com.intellij.ide.IdeEventQueue.ifFocusEventsInTheQueue(IdeEventQueue.kt:187)
	at com.intellij.ide.IdeEventQueue.executeWhenAllFocusEventsLeftTheQueue(IdeEventQueue.kt:153)
	at com.intellij.openapi.wm.impl.FocusManagerImpl.doWhenFocusSettlesDown(FocusManagerImpl.java:168)
	at com.intellij.openapi.wm.impl.FocusManagerImpl.doWhenFocusSettlesDown(FocusManagerImpl.java:162)
	at com.intellij.ui.popup.AbstractPopup.dispose(AbstractPopup.java:1616)
	at com.intellij.ui.popup.WizardPopup.dispose(WizardPopup.java:165)
	at com.intellij.ui.popup.list.ListPopupImpl.dispose(ListPopupImpl.java:391)
	at com.intellij.openapi.util.ObjectTree.runWithTrace(ObjectTree.java:127)
	at com.intellij.openapi.util.ObjectTree.executeAll(ObjectTree.java:159)
	at com.intellij.openapi.util.Disposer.dispose(Disposer.java:264)
	at com.intellij.openapi.util.Disposer.dispose(Disposer.java:252)
	at com.intellij.ui.popup.WizardPopup.disposeAllParents(WizardPopup.java:273)
	at com.intellij.ui.popup.list.ListPopupImpl.disposePopup(ListPopupImpl.java:485)
	at com.intellij.ui.popup.list.ListPopupImpl.handleNextStep(ListPopupImpl.java:512)
	at com.intellij.ui.popup.list.ListPopupImpl._handleSelect(ListPopupImpl.java:479)
	at com.intellij.ui.popup.list.ListPopupImpl.handleSelect(ListPopupImpl.java:426)
	at com.intellij.ui.popup.list.ListPopupImpl$1.actionPerformed(ListPopupImpl.java:281)
	at com.intellij.ui.popup.WizardPopup.proceedKeyEvent(WizardPopup.java:400)
	at com.intellij.ui.popup.WizardPopup.dispatch(WizardPopup.java:372)
	at com.intellij.ui.popup.PopupDispatcher.dispatchKeyEvent(PopupDispatcher.java:115)
	at com.intellij.ui.popup.PopupDispatcher.dispatch(PopupDispatcher.java:157)
	at com.intellij.ide.IdePopupManager.dispatch(IdePopupManager.java:107)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.kt:546)
	at com.intellij.ide.IdeEventQueue.access$_dispatchEvent(IdeEventQueue.kt:68)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1$1.compute(IdeEventQueue.kt:349)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1$1.compute(IdeEventQueue.kt:348)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:787)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1.invoke(IdeEventQueue.kt:348)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1.invoke(IdeEventQueue.kt:343)
	at com.intellij.ide.IdeEventQueueKt.performActivity$lambda$1(IdeEventQueue.kt:994)
	at com.intellij.openapi.application.TransactionGuardImpl.performActivity(TransactionGuardImpl.java:113)
	at com.intellij.ide.IdeEventQueueKt.performActivity(IdeEventQueue.kt:994)
	at com.intellij.ide.IdeEventQueue.dispatchEvent$lambda$4(IdeEventQueue.kt:343)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:831)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.kt:385)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:207)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:128)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:117)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:105)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:92)
```

### PR DESCRIPTION
**Pull request by angelozerr**
fix: java.lang.IndexOutOfBoundsException when trying to "insert @RestClient"

Fixes #764

### CODE REVIEW

**Review by jeffmaury — CHANGES_REQUESTED** [review]
Code does not compile can't test it

**Comment by angelozerr** [comment]
> Code does not compile can't test it

Ooops sorry, I missed to push a java file code, pushed, hope it sould be better.

File: src/main/java/com/redhat/devtools/intellij/lsp4mp4ij/psi/internal/core/java/codeaction/CodeActionHandler.java [hunk 1]
```diff
@@ -195,15 +195,15 @@ public CodeAction resolveCodeAction(CodeAction unresolved, IPsiUtils utils) {
 
 		/*JavaCodeActionResolveContext context = new JavaCodeActionResolveContext(unit, start, end - start, utils, params,
 				unresolved);*/
-		JavaCodeActionResolveContext context = new JavaCodeActionResolveContext(unit.getViewProvider().clone().getPsi(unit.getLanguage()),
+		JavaCodeActionResolveContext context = new JavaCodeActionResolveContext(unit,
```
Comment (adietish) [inline src/main/java/com/redhat/devtools/intellij/lsp4mp4ij/psi/internal/core/java/codeaction/CodeActionHandler.java:198]: @angelozerr can you please explain how these changes fix the issue at hand? Somehow the original code was cloning the view provider, get the psi from if while your new version would use the existing psi and copy the context (see further down). Correct? How does this fix the issue?

File: src/main/java/com/redhat/devtools/intellij/lsp4mp4ij/psi/core/java/codeaction/JavaCodeActionContext.java [hunk 2]
```diff
@@ -1,138 +1,138 @@
...
+ * @author Angelo ZERR
+ *
+ */
+public class JavaCodeActionContext extends AbstractJavaContext implements IInvocationContext {
+
+	private final int selectionOffset;
+	private final int selectionLength;
+
+	private final MicroProfileJavaCodeActionParams params;
+	protected JavaCodeActionContext source;
```
Comment (adietish) [inline src/main/java/com/redhat/devtools/intellij/lsp4mp4ij/psi/core/java/codeaction/JavaCodeActionContext.java:44]: @angelozerr afaics source was changed from `private` to `protected`. That's the only change in this class. The rest are just changes in the formatting. Correct?

**Review by adietish — APPROVED** [review]
this PR fixes my issue, great job! Wondering if we could have a little tests here.

File: src/main/java/com/redhat/devtools/intellij/lsp4mp4ij/psi/core/java/codeaction/JavaCodeActionContext.java [hunk 3]
```diff
@@ -1,138 +1,138 @@
...
-		codeAction.setKind(proposal.getKind());
-		codeAction.setEdit(edit);
-		codeAction.setDiagnostics(Arrays.asList(diagnostics));
-		return codeAction;
-	}
-
-	public WorkspaceEdit convertToWorkspaceEdit(ChangeCorrectionProposal proposal) {
-		WorkspaceEdit edit = ChangeUtil.convertToWorkspaceEdit(proposal.getChange(), getUri(), getUtils(),
-				params.isResourceOperationSupported());
-		if (!ChangeUtil.hasChanges(edit)) {
```
Comment (adietish) [inline src/main/java/com/redhat/devtools/intellij/lsp4mp4ij/psi/core/java/codeaction/JavaCodeActionContext.java:132]: ex. a test for `ChangeUtil` could assert the 3 cases present in #hasChanges. Would this make sense?

**Pull request merged**